### PR TITLE
Merge pending changes into master

### LINKED ISSUE #246: [Feature Request] Area command/config to prevent users from shouting in an area.
Similar to /allow_blankposting, but for shouting rather than blankposting. Members with elevated perms (staff, and maybe CM) would still be able to shout. Useful for lobby areas or chatrooms where shouts are just loud and obnoxious or when things are being spammed and need to be cooled down.

### LINKED ISSUE #257: Spectators can play music
Whether spectators shouldn't be allowed to change the music, or it should be displaying it properly that they did, either way there's an issue here. Presumably it breaks trying to find the character name/showname of the spectator as they have none and displays nothing as a result, but it still does change the music, which looks quite odd.

### LINKED ISSUE #264: When two people send a message at the same time in different areas, one gets blocked from sending.
It exhibits similar behaviour as though they were in the same area, and by blocked, I mean the user's message remains in their textbox as though they didn't hit enter.

### PR DESCRIPTION
**Pull request by Salanto**
*Refactors Websocket implementation on Akashi
*Enables per-Area WTCE and Shout control
*Upgrades Akashi to a role-based permission system
*Disables spectators from playing music
*Reworks rate limit for IC messages
*Privatises Server for the most part
*Reworks AOPacket to be a bit cleaner
*Enforces clang-format on the CI and providers QtCreator clang-format file
*Update repository readme
*Enforce handshake checks
*Easier command alias and permission extension feature.
* Fixes bug where localhost may not be detected correctly.
* maybe fix QIODEVICE open error
* Update DB to version 3 and prevent root permission lockout scenario on upgrade

closes #264 
closes #257 
closes #246

### CODE REVIEW

**Comment by in1tiate** [comment]
Oh dear.

**Review by in1tiate — CHANGES_REQUESTED** [review]
scary diff but most of it is formatting. got some stuff i'd like corrected before i merge this but overall it's really solid

File: core/src/aoclient.cpp [hunk 1]
```diff
@@ -17,55 +17,154 @@
...
-    }
-}
+#include "include/area_data.h"
+#include "include/command_extension.h"
+#include "include/config_manager.h"
+#include "include/db_manager.h"
+#include "include/network/aopacket.h"
+#include "include/server.h"
+
+const QMap<QString, AOClient::CommandInfo> AOClient::COMMANDS{
```
[the remark this answers is not in the document]
Comment (Salanto) [reply core/src/aoclient.cpp:27]: Not quite able to do this as we need the entire command table on the stack to allocate our alias stuff.

**Comment by Salanto** [comment]
closes #166 
The original problem seems to have been improper parenting causing edge cases where the client was deleted, but not its corresponding socket.

**Review by in1tiate — APPROVED** [review]
Very, very solid work here. Thank you to all who contributed.

**Pull request merged**